### ISSUE TITLE
libappimage 0.2 tests code coverage

### ISSUE THREAD

**Issue by azubieta** [AppImage/libappimage#39]
In order to have a reliable code base and to not jump blindly into a release, `libappimage` 0.2 must have an overall code coverage indicator of 90%. This issue is to keep @probonopd and @TheAssassin aware of the progress.

**Comment by azubieta** [comment]
Current status, newly written code is in the `src/libappimage/core`, `src/libappimage/core/impl` and `src/libappimage/utils`.

Inmediate goal raise those three to **90%** or more.
![screenshot_20190103_140245](https://user-images.githubusercontent.com/1138094/50658855-44f9a800-0f60-11e9-8f63-c5f8738390ad.png)

**Comment by TheAssassin** [comment]
It'd be great if you could measure the unit test coverage separately, too.

**Comment by azubieta** [comment]
The issue with the unit-tests and the AppImages is that we need a whole file test a single function. Which is against the idea of testing things in isolation. In the case of the utility classes is totally doable.

**Comment by TheAssassin** [comment]
I disagree with this point of view. It depends on how you write the code. For instance, a method that parses the magic bytes from an AppImage could be written to accept an `istream` instead of a path. Then, you can pass a `stringstream` or so within the scope of a unit test. That'll cover 95% of the code.

I don't know all the new code, so there might be cases in which such tests can't be written. But in my experience, it's normally possible to structure code in a way so that it's testable from unit tests.

**Comment by azubieta** [comment]
After implementing some tests to libappimage/utils/Elf we these results. We are missing the cases were the AppImage payload is partially broken and will produce errors to read a given file or to go to the next entry. I have been trying to randomly override the AppImage file in order to break it but without much success.

Ideas of how to create such partially brocken payloads are wellcome.
![screenshot_20190103_155421](https://user-images.githubusercontent.com/1138094/50663612-dae8ff00-0f6f-11e9-949c-c6f9f7f6d843.png)

**Comment by TheAssassin** [comment]
Probably not a good idea as it'll generate a lot of code, but have you ever thought about using the https://en.wikipedia.org/wiki/Template_method_pattern to allow for breaking up algorithms into components? Via subclassing, one could exchange annoying parts with stubs, and test the rest of the functionality.

**Comment by azubieta** [comment]
That's pretty much what the [Traversal](https://github.com/AppImage/libappimage/blob/rewrite_in_cpp/src/libappimage/core/Traversal.h) class and its derivatives are.

**Comment by azubieta** [comment]
Update on the code coverage of tests cases. Right now our major ofenders are the legacy code and the error management code sections.
![screenshot_20190204_121724](https://user-images.githubusercontent.com/1138094/52228029-e3866b00-2876-11e9-9e2c-4a84a30e3b19.png)

**Comment by azubieta** [comment]
Already found the way of publishing our tests code coverage results to codacy. @probonopd if you have an organization created there would you mind to add me.

**Comment by TheAssassin** [comment]
I still prefer textual reports without the need for any other annoying service.

**Comment by azubieta** [comment]
This won't exclude textual reports. It takes the `gcovr` results and publishes them to www.codacy.com.

**Comment by TheAssassin** [comment]
Again: I want to see textual reports in the Travis logs. I don't want to prevent you from using any crap services like Codacy. Just set up an alternative for people who value their own data.
